File: matlab_code/w2l.m
function L=w2l(W, idx_closest)
% W2L weight matrix to Laplacian matrix
% 
% L=w2l(W)
% 
% by: KS Lu
% 20170712
%
sz_W = size(W);
C = zeros(sz_W);
if any(W<0)
    %error('W is not a valid weight matrix');
end

if (nargin > 1)
    C(idx_closest,idx_closest) = 2;
    L = C + diag(sum(W))-W+diag(diag(W));
else
    L = diag(sum(W))-W+diag(diag(W));
end


end

File: python_translation/w2l.py
import numpy as np

def w2l(W, idx_closest=None):
    """
    Convert weight matrix to Laplacian matrix.

    Args:
        W (numpy.ndarray): Weight matrix.
        idx_closest (numpy.ndarray): Indices of the closest points (optional).

    Returns:
        L (numpy.ndarray): Laplacian matrix.
    """
    sz_W = W.shape
    C = np.zeros(sz_W)

    if np.any(W < 0):
        # raise ValueError('W is not a valid weight matrix')
        pass  # Handle negative weights differently if needed

    if idx_closest is not None:
        C[np.ix_(idx_closest, idx_closest)] = 2

    L = C + np.diag(np.sum(W, axis=1)) - W + np.diag(np.diag(W))

    return L
